### PR TITLE
[dev] Basic simplebus feature.

### PR DESCRIPTION
**Pull request by MichalBlk**
This is an initial approach to implement the simplebus bus. The first step is to stop assigning resources explicitly and parse the FDT instead. We still need to implement the simplebus enumeration but that will be introduced in a separate PR after the RISC-V port is finally merged into the master branch.

### CODE REVIEW

**Comment by cahirwpz** [comment]
@pj1031999 If you started working on the review of this PR, please do not continue your work. We need you to give feedback on more important PRs like #1230

File: include/dev/simplebus.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#ifndef _DEV_SIMPLEBUS_H_
+#define _DEV_SIMPLEBUS_H_
+
+typedef uint32_t phandle_t;
+typedef struct device device_t;
```
Comment (pj1031999) [inline include/dev/simplebus.h:5]: Let's include real source of these typedefs.

File: sys/drv/simplebus.c [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+  int icells;
+  int err = FDT_intr_cells(iparent, &icells);
+  if (err)
+    return err;
+
+  pcell_t *tuples;
+  int ntuples =
+    FDT_getencprop_alloc_multi(node, "interrupts", icells, (void **)&tuples);
+  if (ntuples < 0)
```
Comment (pj1031999) [inline sys/drv/simplebus.c:58]: Error is only on `-1`. Am I missing something?

File: sys/drv/simplebus.c [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+static int sb_get_interrupts_extended(device_t *dev, fdt_intr_t *intrs,
+                                      size_t *cntp) {
+  phandle_t node = dev->node;
+
+  pcell_t *cells;
+  int ncells = FDT_getencprop_alloc_multi(node, "interrupts-extended",
+                                          sizeof(pcell_t), (void **)&cells);
+  if (ncells < 0)
```
Comment (pj1031999) [inline sys/drv/simplebus.c:86]: ditto

File: sys/kern/fdt.c [hunk 4]
```diff
@@ -130,14 +144,58 @@ ssize_t FDT_getencprop(phandle_t node, const char *propname, pcell_t *buf,
   if (buflen % 4)
     return -1;
 
-  ssize_t ret = FDT_getprop(node, propname, buf, buflen);
-  if (ret < 0)
+  ssize_t rv = FDT_getprop(node, propname, buf, buflen);
+  if (rv < 0)
```
Comment (pj1031999) [inline sys/kern/fdt.c:148]: Please avoid name changes if you introduce new features and do that in separate commits.

File: sys/kern/fdt.c [hunk 5]
```diff
@@ -130,14 +144,58 @@ ssize_t FDT_getencprop(phandle_t node, const char *propname, pcell_t *buf,
...
+      FDT_free(buf);
+      buf = NULL;
+      goto end;
+    }
+  }
+  rv = len / elsz;
+
+end:
+  *bufp = buf;
+  return rv;
```
Comment (pj1031999) [inline sys/kern/fdt.c:177]: You can just return `-1` and leave `*bufp` unchanged in case of error. For me it will be more readable.

File: include/dev/simplebus.h (hunk 1 above)
Comment (MichalBlk) [reply include/dev/simplebus.h]: Done.

File: sys/drv/simplebus.c (hunk 2 above)
Comment (MichalBlk) [reply sys/drv/simplebus.c]: You're right. I changed it to ` == -1`.

File: sys/drv/simplebus.c (hunk 3 above)
Comment (MichalBlk) [reply sys/drv/simplebus.c]: Done.

File: sys/kern/fdt.c (hunk 4 above)
Comment (MichalBlk) [reply sys/kern/fdt.c]: Good point. I removed these changes from this PR.

File: sys/kern/fdt.c (hunk 5 above)
Comment (MichalBlk) [reply sys/kern/fdt.c]: I agree. It makes the code much more readable. Done.

**Review by cahirwpz — APPROVED** [review]
